pkg/controller/volume/selinuxwarning/cache/volumecache.go

```diff
@@ -0,0 +1,260 @@
+// Remove a pod from the cache. Prunes all empty structures.
+func (c *volumeCache) DeletePod(logger klog.Logger, podKey cache.ObjectName) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	defer c.dump(logger)
+
+	for volumeName, volume := range c.volumes {
+		delete(volume.pods, podKey)
+		if len(volume.pods) == 0 {
+			delete(c.volumes, volumeName)
+		}
+	}
+}
+
+func (c *volumeCache) dump(logger klog.Logger) {
+	if !logger.V(dumpLogLevel).Enabled() {
+		return
+	}
+	logger.Info("VolumeCache dump:")
+
+	// sort the volume to have consistent output
+	volumeIDs := make([]v1.UniqueVolumeName, 0, len(c.volumes))
+	for volumeID := range c.volumes {
+		volumeIDs = append(volumeIDs, volumeID)
+	}
+	sort.Slice(volumeIDs, func(i, j int) bool {
+		return volumeIDs[i] < volumeIDs[j]
+	})
+	for _, volumeID := range volumeIDs {
+		volume := c.volumes[volumeID]
+		logger.Info("Cached volume", "volume", volumeID, "csiDriver", volume.csiDriver)
+
+		// Sort the pods to have consistent output
+		podKeys := make([]cache.ObjectName, 0, len(volume.pods))
+		for podKey := range volume.pods {
+			podKeys = append(podKeys, podKey)
+		}
+		sort.Slice(podKeys, func(i, j int) bool {
+			return podKeys[i].String() < podKeys[j].String()
+		})
+		for _, podKey := range podKeys {
+			podInfo := volume.pods[podKey]
+			logger.Info("  pod", "pod", podKey, "label", podInfo.label, "changePolicy", podInfo.changePolicy)
+		}
+	}
+}
+
+// GetPodsForCSIDriver returns all pods that use volumes with the given CSI driver.
+func (c *volumeCache) GetPodsForCSIDriver(driverName string) []cache.ObjectName {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	var pods []cache.ObjectName
+	for _, volume := range c.volumes {
+		if volume.csiDriver != driverName {
+			continue
+		}
+		for podKey := range volume.pods {
+			pods = append(pods, podKey)
+		}
+	}
+	return pods
+}
+
+// SendConflicts sends all current conflicts to the given channel.
+func (c *volumeCache) SendConflicts(logger klog.Logger, ch chan<- Conflict) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	logger.V(4).Info("Scraping conflicts")
+	c.dump(logger)
+
+	for _, volume := range c.volumes {
+		// compare pods that use the same volume with each other
+		for podKey, podInfo := range volume.pods {
+			for otherPodKey, otherPodInfo := range volume.pods {
+				if podKey == otherPodKey {
+					continue
+				}
+				// create conflict only for the first pod. The other pod will get the same conflict in its own iteration of `volume.pods` loop.
+				if podInfo.changePolicy != otherPodInfo.changePolicy {
+					ch <- Conflict{
+						PropertyName:       "SELinuxChangePolicy",
+						EventReason:        "SELinuxChangePolicyConflict",
+						Pod:                podKey,
+						PropertyValue:      string(podInfo.changePolicy),
+						OtherPod:           otherPodKey,
+						OtherPropertyValue: string(otherPodInfo.changePolicy),
+					}
+				}
+				if podInfo.label != otherPodInfo.label {
+					ch <- Conflict{
+						PropertyName:       "SELinux label",
```

Comment:
Since we are reporting this as a metric dimension, I am not sure keeping the whitespace is a good idea.

Reply:
Changed to `SELinuxLabel` + updated the field description.